File: BigO Blue/code/1. Dynamic Array/DynamicArray.cpp
// Dynamic Array (mảng động) là cấu trúc dữ liệu dùng để lưu trữ các đối tượng có kích thước không xác định.
// Khác với mảng tĩnh, người dùng không cần phải khai báo trước số lượng phần tử khi sử dụng.

#include <iostream>
#include <vector>

using namespace std;

int main(){
    vector<int> v;

    // thêm phần tử
    v.push_back(5);
    v.push_back(7);
    v.push_back(8);
    v.push_back(3);
    v.push_back(6);

    // chèn giá trị -- insert (iterator, val)
    vector<int>::iterator it;
    it = v.begin() + 2;
    v.insert(it, 9);

    // lấy kích thước -- size()
    int n = v.size();
    cout << n << "\n";

    // truy cập phần tử -- using []
    int result = v[2];
    cout << result << "\n";

    // truy cập để thay đổi giá trị -- using []
    v[4] = 9;

    // xoá phần tử cuối -- pop_back()
    v.pop_back();

    // xoá phần tử ở vị trí bất kỳ -- erase(iterator)
    vector<int>::iterator a;
    a = v.begin() + 2;
    v.erase(a);

    // xoá toàn bộ các phần tử trong dynamic array -- clear()
    v.clear();

    // tăng kích thước của dynamic array -- resize()
    v.resize(7);

    // giảm kích thước của dynamic array -- resize()
    v.resize(2);
    
    // kiểm tra xem dynamic array có rỗng không -- empty()
    if (v.empty() == true)
        cout << "DA is empty";
    else
        cout << "DA is not empty";

    // traverse
    // xuôi
    for (int i = 0; i < v.size(); i++){
        cout << v[i] << ", ";
    }

    vector<int>::iterator ik;
    for(ik = v.begin(); ik != v.end(); ik++){
        cout << *ik << ", ";
    }

    // ngược
    for (int i = v.size() - 1; i >= 0; i--){
        cout << v[i] << ", ";
    }

    vector<int>::reverse_iterator j;
    for (j = v.rbegin(); j != v.rend(); j++){
        cout << *j << ", ";
    }


}

File: BigO Blue/code/1. Dynamic Array/DynamicArray.py
# Dynamic Array (Mảng động) là cấu trức dữ liệu dùng để lưu trữ các đối
# tượng có kích thước không xác định. Khác với mang tĩnh, người dùng
# không cần phải khai báo trước số lượng phần tử khi sử dụng.

# Khởi tạo dynamic array
variable = []

# thêm phần tử
variable.append(5)
variable.append(7)
variable.append(8)
variable.append(3)
variable.append(6)

# chèn giá trị vào dynamic array -- insert(pos, obj)
variable.insert(2, 9)

# lấy kích thước của dymnamic array -- len(obj)
n = len(variable)

print(n)

# truy cập phần tử trong dynamic array -- using []
result = variable[2]

print(result)

# xoá phần tử cuối khỏi dynamic array -- pop()
variable.pop()

# xoá phần tử ở vị trí bất kỳ trong dynamic array -- pop(pos)
variable.pop(2)

# xoá toàn bộ phần tử trong dynamic array -- clear()
variable.clear()

# tăng kích thước của dynamic array (ta nối thêm list mới vào list hiện tại)  -- extend(list)
variable.extend(2*[0])

# giảm kích thước của dynamic array -- using slicing
variable = variable[0: 2]

# traverse
# xuôi
for i in range(len(variable)):
    print(variable[i], end = ', ')

# ngược
for i in range(len(variable) - 1, -1, -1):
    print(variable[i], end = ', ')


